File: redirect-service/src/main/java/com/encurtador/redirect_service/config/RabbitConfig.java
package com.encurtador.redirect_service.config;

import com.rabbitmq.client.Connection;
import com.rabbitmq.client.ConnectionFactory;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Profile;
import reactor.core.publisher.Mono;
import reactor.core.scheduler.Schedulers;
import reactor.rabbitmq.RabbitFlux;
import reactor.rabbitmq.Sender;
import reactor.rabbitmq.SenderOptions;
import reactor.util.retry.Retry;

import java.time.Duration;

@Configuration
@Slf4j
@Profile("!test")
public class RabbitConfig {
    @Value("${spring.rabbitmq.host}") private String host;
    @Value("${spring.rabbitmq.port}") private int port;
    @Value("${spring.rabbitmq.username}") private String username;
    @Value("${spring.rabbitmq.password}") private String password;
    @Value("${spring.rabbitmq.virtual-host}") private String virtualHost;

    @Bean
    Mono<Connection> connectionMono() {
        ConnectionFactory connectionFactory = new ConnectionFactory();
        connectionFactory.setHost(host);
        connectionFactory.setPort(port);
        connectionFactory.setUsername(username);
        connectionFactory.setPassword(password);
        connectionFactory.setVirtualHost(virtualHost);
        connectionFactory.useNio();
        return Mono.fromCallable(() -> connectionFactory.newConnection("reactor-rabbit"))
                .subscribeOn(Schedulers.boundedElastic())
                .cache()
                .doOnSubscribe(sub -> log.info("Establishing RabbitMQ connection..."))
                .doOnError(err -> log.error("Failed to establish RabbitMQ connection", err))
                .doOnSuccess(conn -> log.info("RabbitMQ connection established successfully"))
                .retryWhen(Retry.backoff(10, Duration.ofSeconds(1))
                        .maxBackoff(Duration.ofSeconds(30))
                        .doBeforeRetry(rs -> log.warn("Retrying RabbitMQ connection... Attempt: {}", rs.totalRetries() + 1))
                );
    }

    @Bean
    public SenderOptions senderOptions(Mono<Connection> connectionMono) {
        return new SenderOptions()
                .connectionMono(connectionMono)
                .resourceManagementScheduler(Schedulers.boundedElastic())
                .connectionSubscriptionScheduler(Schedulers.boundedElastic());
    }

    @Bean
    public Sender sender(SenderOptions senderOptions) {
        return RabbitFlux.createSender(senderOptions);
    }
}

File: short-link-metrics-service/src/main/kotlin/com/encurtaai/short_link_metrics_service/config/RabbitConfig.kt
package com.encurtaai.short_link_metrics_service.config

import com.rabbitmq.client.Connection
import com.rabbitmq.client.ConnectionFactory
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import reactor.core.publisher.Mono
import reactor.core.scheduler.Schedulers
import reactor.rabbitmq.RabbitFlux
import reactor.rabbitmq.Receiver
import reactor.rabbitmq.ReceiverOptions
import reactor.util.retry.Retry
import java.time.Duration

@Configuration
class RabbitConfig(
    @Value("\${spring.rabbitmq.host}") private val host: String,
    @Value("\${spring.rabbitmq.port}") private val port: Int,
    @Value("\${spring.rabbitmq.username}") private val username: String,
    @Value("\${spring.rabbitmq.password}") private val password: String,
    @Value("\${spring.rabbitmq.virtual-host}") private val virtualHost: String
) {
    private val log = LoggerFactory.getLogger(RabbitConfig::class.java)

    @Bean
    fun connectionMono(): Mono<Connection> {
        val factory = ConnectionFactory()
        factory.host = host
        factory.port = port
        factory.username = username
        factory.password = password
        factory.virtualHost = virtualHost
        factory.useNio()
        return Mono.fromCallable { factory.newConnection("reactor-rabbit") }
            .subscribeOn(Schedulers.boundedElastic())
            .cache()
            .doOnSubscribe { log.info("Attempting to connect to RabbitMQ...") }
            .doOnError { ex -> log.error("Error connecting to RabbitMQ: ${ex.message}", ex) }
            .doOnSuccess { log.info("Successfully connected to RabbitMQ") }
            .retryWhen(
                Retry.backoff(10, Duration.ofSeconds(1))
                    .maxBackoff(Duration.ofSeconds(30))
                    .doBeforeRetry { rs ->
                        log.warn("Failed to connect to RabbitMQ. Retry #${rs.totalRetries() + 1}")
                    }
            )
    }

    @Bean
    fun receiverOptions(connectionMono: Mono<Connection>): ReceiverOptions =
        ReceiverOptions()
            .connectionMono(connectionMono)
            .connectionSubscriptionScheduler(Schedulers.boundedElastic())

    @Bean
    fun receiver(receiverOptions: ReceiverOptions): Receiver =
        RabbitFlux.createReceiver(receiverOptions)
}
